mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DefaultDaoReturnTypeParser.java

```diff
@@ -0,0 +1,196 @@
+          .put(Row.class, DaoReturnType.ROW)
+          .put(ResultSet.class, DaoReturnType.RESULT_SET)
+          .build();
+
+  /**
+   * The return types that correspond to a generic class with a single type parameter that is an
+   * entity class.
+   */
+  protected static final Map<Class<?>, DaoReturnTypeKind> DEFAULT_ENTITY_CONTAINER_MATCHES =
+      ImmutableMap.<Class<?>, DaoReturnTypeKind>builder()
+          .put(Optional.class, DefaultDaoReturnTypeKind.OPTIONAL_ENTITY)
+          .put(CompletionStage.class, DefaultDaoReturnTypeKind.FUTURE_OF_ENTITY)
+          .put(CompletableFuture.class, DefaultDaoReturnTypeKind.FUTURE_OF_ENTITY)
+          .put(PagingIterable.class, DefaultDaoReturnTypeKind.PAGING_ITERABLE)
+          .build();
+
+  /** The return types that correspond to a future of a non-generic Java class. */
+  protected static final Map<Class<?>, DaoReturnType> DEFAULT_FUTURE_OF_CLASS_MATCHES =
+      ImmutableMap.<Class<?>, DaoReturnType>builder()
+          .put(Void.class, DaoReturnType.FUTURE_OF_VOID)
+          .put(Boolean.class, DaoReturnType.FUTURE_OF_BOOLEAN)
+          .put(Long.class, DaoReturnType.FUTURE_OF_LONG)
+          .put(Row.class, DaoReturnType.FUTURE_OF_ROW)
+          .put(AsyncResultSet.class, DaoReturnType.FUTURE_OF_ASYNC_RESULT_SET)
+          .build();
+
+  /**
+   * The return types that correspond to a future of a generic class with a single type parameter
+   * that is an entity class.
+   */
+  protected static final Map<Class<?>, DaoReturnTypeKind>
+      DEFAULT_FUTURE_OF_ENTITY_CONTAINER_MATCHES =
+          ImmutableMap.<Class<?>, DaoReturnTypeKind>builder()
+              .put(Optional.class, DefaultDaoReturnTypeKind.FUTURE_OF_OPTIONAL_ENTITY)
+              .put(
+                  MappedAsyncPagingIterable.class,
+                  DefaultDaoReturnTypeKind.FUTURE_OF_ASYNC_PAGING_ITERABLE)
+              .build();
+
+  protected final ProcessorContext context;
+  private final Map<TypeKind, DaoReturnType> typeKindMatches;
+  private final Map<Class<?>, DaoReturnType> classMatches;
+  private final Map<Class<?>, DaoReturnTypeKind> entityContainerMatches;
+  private final Map<Class<?>, DaoReturnType> futureOfClassMatches;
+  private final Map<Class<?>, DaoReturnTypeKind> futureOfEntityContainerMatches;
+
+  public DefaultDaoReturnTypeParser(ProcessorContext context) {
+    this(
+        context,
+        DEFAULT_TYPE_KIND_MATCHES,
+        DEFAULT_CLASS_MATCHES,
+        DEFAULT_ENTITY_CONTAINER_MATCHES,
+        DEFAULT_FUTURE_OF_CLASS_MATCHES,
+        DEFAULT_FUTURE_OF_ENTITY_CONTAINER_MATCHES);
+  }
+
+  protected DefaultDaoReturnTypeParser(
+      ProcessorContext context,
+      Map<TypeKind, DaoReturnType> typeKindMatches,
+      Map<Class<?>, DaoReturnType> classMatches,
+      Map<Class<?>, DaoReturnTypeKind> entityContainerMatches,
+      Map<Class<?>, DaoReturnType> futureOfClassMatches,
+      Map<Class<?>, DaoReturnTypeKind> futureOfEntityContainerMatches) {
```

Comment:
OK, this looks much better now. Subclasses can use this constructor to pass their own maps instead of the defaults, so it would be trivial to support other container types (e.g. `Publisher<Product>`) without copy-pasting everything.

This also makes `parse()` easier to read I think.